tools/test

```diff
@@ -117,6 +117,10 @@ run_native() {
     )
 }
 
+run_native() {
+    run_native_android || return
```

Comment:
Huh, interesting! It looks like something strange is happening.

I added `|| return` because I found that `run_native` wasn't exiting when `run_native_android` failed.

I've tried a few things just now, and I think I found the culprit. From the Bash manual on `set -e` (emphasis mine),

> The shell does not exit if the command that fails is part of the command list immediately following a `while` or `until` keyword, part of the test in an `if` statement, _**part of any command executed in a `&&` or `||` list except the command following the final `&&` or `||`**_, any command in a pipeline but the last, or if the command’s return status is being inverted with `!`.

If I make this change just after the `for suite in "${suites[@]}"` loop, it works as expected with the `|| return` removed:

```diff
-    esac || failed+=($suite)
+    esac
```

🤯

Reply:
Does that sound right, and if so, how would you like to proceed?

Reply:
Urgh, right. Thanks for catching that. This is the main gotcha in `set -e`, where it doesn't do everything you'd think it does (because any reasonable version of it absolutely would.) :exploding_head: is right on.

This also explains why I wrote a similar `|| return` in one spot in the existing code, at `check_node || return`. When I wrote this comment I was wondering about that...

I guess the short of it is that inside all these functions, we effectively *don't* have the benefit of `set -e`. And we can't even re-enable it from inside each function, as this bit later in that same manual item explains:
> If a compound command or shell function sets `-e` while executing in a context where `-e` is ignored, that setting will not have any effect until the compound command or the command containing the function call completes.

I think the right immediate fix is to use `|| return` just as this branch does. Basically we need that wherever we have a nontrivial command that isn't the final command of its function. I think the existing code is all OK in this respect, thanks to that `|| return` in one place we did need it.

That and maybe a comment on the `set -e` at the top, warning that it doesn't (and sadly can't) actually apply to most of the code.

Then... if we want to get back `set -e` behavior inside these functions, while still being able to "catch" its propagation at the outer loop so that we can go on to run the other suites, I think basically what's required is to fork a subshell (with the `( … )` syntax -- like we already do inside `run_android` in order to use `cd` and contain its effect.) I think that's the real point of what that quoted bit of docs is saying: `set -e` stops having any effect if it isn't going to be able to go all the way to exit the whole shell. That's a bit of overhead, but only like a millisecond each:
```
$ time bash -c 'for i in {1..100}; do ( : ); done'

time: 0.099s wall (0.065s u, 0.041s s)
```
so not a problem at all. (That fast startup time is one of Bash's handful of great strengths, compared to almost any other language that doesn't require a compile step.)

Reply:
OK, sounds good! Glad to see the 🤯 feeling is shared! 😅 I'll put a comment on the `set -e`.

Reply:
I'm not sure what's the best strategy for forking a subshell where appropriate; we'd of course like to leave a good signpost if future added `run_foo` functions have to remember to do something.